### PR TITLE
8294847: Fix calculation of G1 effective scanned cards prediction

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews of this change that fixes the prediction of the actually to be scanned cards?

In particular, the current code completely misses the log buffer cards in the "effective" number of scanned cards. The change simply assumes that all pending log buffer cards are going to be scanned, ignoring duplicates.

Other changes are just removing related, unused or unnecessary code.

Depends on PR#10644, so please review that one first.

Testing: gha, local testing, tier1-5 with other changes in this patch series, perf testing with other changes in this patch series

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294847](https://bugs.openjdk.org/browse/JDK-8294847): Fix calculation of G1 effective scanned cards prediction

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10647/head:pull/10647` \
`$ git checkout pull/10647`

Update a local copy of the PR: \
`$ git checkout pull/10647` \
`$ git pull https://git.openjdk.org/jdk pull/10647/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10647`

View PR using the GUI difftool: \
`$ git pr show -t 10647`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10647.diff">https://git.openjdk.org/jdk/pull/10647.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1Analytics.hpp [hunk 1]
```diff
@@ -139,10 +139,9 @@ class G1Analytics: public CHeapObj<mtGC> {
 
   double predict_concurrent_refine_rate_ms() const;
   double predict_dirtied_cards_rate_ms() const;
-  double predict_young_card_merge_to_scan_ratio() const;
-
-  double predict_mixed_card_merge_to_scan_ratio() const;
 
+  // Predict how many of the given remembered set of length rs_length will add to
+  // the number of total cards scanned.
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Analytics.hpp:144]: I'm having a very hard time parsing that comment.  I _think_ it is equivalent to "Predict how many cards in a remembered set of length rs_length will need to be scanned."

File: src/hotspot/share/gc/g1/g1Policy.cpp [hunk 2]
```diff
@@ -1010,7 +1010,9 @@ void G1Policy::record_young_gc_pause_end(bool evacuation_failed) {
 
 double G1Policy::predict_base_time_ms(size_t pending_cards,
                                       size_t rs_length) const {
-  size_t effective_scanned_cards = _analytics->predict_scan_card_num(rs_length, collector_state()->in_young_only_phase());
+  // Assume that all cards from the log buffers will be scanned, i.e. there are no
+  // duplicates in that set.
+  size_t effective_scanned_cards = _analytics->predict_scan_card_num(rs_length, collector_state()->in_young_only_phase()) + pending_cards;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Policy.cpp:1015]: Consider breaking up this very long line.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1Analytics.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1Analytics.hpp]: This predictor is a bit complicated: it predicts for that given remembered set length, what part of it adds to the number of scanned cards - assuming that the pending log buffers are always scanned, we need to find out how many additional cards the remembered set will dirty (and need to be scanned). I.e. the remembered set cards that do not duplicate the log buffer cards.
I tried to reformulate that, please let me know if it is more understandable now.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by tschatzl** [comment]
Thanks @kimbarrett @albertnetymk for your reviews.

/integrate

**Pull request closed without merging**